### PR TITLE
feat(tests): implement initial functional test

### PR DESCRIPTION
**Pull request by arein**
# Description

Implements an initial functional test just covering `/health`.

Main purpose is to set this up for actual tests.
Currently requires a local postgres instance to run.

Towards #13 

## How Has This Been Tested?

`cargo test`

## Due Diligence

* [ ] Breaking change
* [ ] Requires a documentation update
* [ ] Requires a e2e/integration test update

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -52,6 +52,9 @@ chrono = "0.4"
 
 [dev-dependencies]
 serial_test = "0.9"
+test-context = "0.1"
+anyhow = "1"
```
Comment (HarryET) [inline Cargo.toml]: Please create an error enum and use `thiserror` as done in `./src/error.js`

File: src/lib.rs [hunk 2]
```diff
@@ -37,11 +38,12 @@ pub mod relay;
 pub mod state;
 pub mod stores;
 
-pub async fn bootstap(config: Config) -> error::Result<()> {
+pub async fn bootstap(mut shutdown: broadcast::Receiver<()>, config: Config) -> error::Result<()> {
     // Check config is valid and then throw the error if its not
     config.is_valid()?;
 
     let pg_options = PgConnectOptions::from_str(&config.database_url)?
+        .username("postgres".into())
```
Comment (HarryET) [inline src/lib.rs]: Not needed, please just include in the database url

File: tests/context/server.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+        let (signal, shutdown) = broadcast::channel(1);
+
+        std::thread::spawn(move || {
+            rt.block_on(async move {
+                let config: Config = Config {
+                    port: public_port,
+                    log_level: "INFO".into(),
+                    relay_url: "https://relay.walletconnect.com".into(),
+                    database_url: "postgres://localhost:5432/postgres".into(),
```
Comment (HarryET) [inline tests/context/server.rs]:
```suggestion
                    database_url: "postgres://user:pass@localhost:5432/postgres".into(),
```
This stops the need for the `.username()` as per the comment above

**Review by HarryET — CHANGES_REQUESTED** [review]

**Comment by HarryET** [comment]
Please also look at CI runs @arein as it appears your tests aren't running and panic

**Comment by arein** [comment]
> Details

Yup tests now require a DB. Will add in follow up PR. Tests are already broken so not making things work. But one of the top prios of course for follow up PR.

**Comment by arein** [comment]
> > Details
> 
> Yup tests now require a DB. Will add in follow up PR. Tests are already broken so not making things work. But one of the top prios of course for follow up PR.

Added in latest revision. Tests are now passing

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -105,6 +105,11 @@ jobs:
           sccache --stop-server || true
           sccache --start-server
 
+      - name: "Run postgres"
```
Comment (HarryET) [inline .github/workflows/ci.yml]: Please use github actions service containers, https://docs.github.com/en/actions/using-containerized-services/about-service-containers

File: tests/context/server.rs [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+async fn wait_for_server_to_shutdown(port: u16) -> error::Result<()> {
+    let poll_fut = async {
+        while !is_port_available(port) {
+            sleep(Duration::from_millis(10)).await;
+        }
+    };
+
+    tokio::time::timeout(Duration::from_secs(3), poll_fut)
+        .map_err(|_| {
```
Comment (HarryET) [inline tests/context/server.rs]: Might be worth using a new Error enum specifically for tests, also have a transparent error so you can use the `?` syntax

**Review by HarryET — CHANGES_REQUESTED** [review]

File: tests/context/server.rs [hunk 6]
```diff
@@ -33,7 +33,7 @@ impl EchoServer {
                     port: public_port,
                     log_level: "INFO".into(),
                     relay_url: "https://relay.walletconnect.com".into(),
-                    database_url: "postgres://postgres@localhost:5432/postgres".into(),
+                    database_url: "postgres://postgres:postgres@localhost:5432/postgres".into(),
```
Comment (HarryET) [inline tests/context/server.rs]: Why not pass this as an env?

Comment (arein) [reply tests/context/server.rs]: Just hacking around for now. Think I found the issue

Comment (arein) [reply tests/context/server.rs]: Cant test this locally so that's why hacking up here :)

**Review by HarryET — APPROVED** [review]
LGTM, sorry for the nitpicky reviews

**Pull request merged**